core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/impl/QueryJoinOptimizer.java

```diff
@@ -236,11 +235,11 @@ protected <M extends Map<Var, Integer>> M getVarFreqMap(List<Var> varList, M var
 			return varFreqMap;
 		}
 
-		protected List<Extension> getExtensions(List<TupleExpr> expressions) {
-			List<Extension> extensions = new ArrayList<>();
+		protected List<TupleExpr> getExtensions(List<TupleExpr> expressions) {
+			List<TupleExpr> extensions = new ArrayList<>();
 			for (TupleExpr expr : expressions) {
-				if (expr instanceof Extension) {
-					extensions.add((Extension) expr);
+				if (TupleExprs.containsExtension(expr)) {
+					extensions.add(expr);
```

Comment:
This is the core of the fix. The problem was that the join argument is a filter condition which in turn  has the Extension as an argument. This is a bit obfuscated because later on the FilterOptimizer removes the filter condition altogether. The fix ensures that we also look deeper down in the algebra tree, so that any wrapped Extensions are also treated as priority arguments.